kaldb/src/main/java/com/slack/kaldb/chunkManager/ChunkManagerBase.java

```diff
@@ -79,42 +92,67 @@ public SearchResult<T> query(SearchQuery query, Duration queryTimeout) {
-                        .orTimeout(queryTimeout.toMillis(), TimeUnit.MILLISECONDS))
-            .map(
-                chunkFuture ->
-                    chunkFuture.exceptionally(
-                        err -> {
-                          LOG.warn("Chunk Query Exception: ", err);
-                          // We catch IllegalArgumentException ( and any other exception that
-                          // represents a parse failure ) and instead of returning an empty result
-                          // we throw back an error to the user
-                          if (err.getCause() instanceof IllegalArgumentException) {
-                            throw (IllegalArgumentException) err.getCause();
-                          }
-                          return errorResult;
-                        }))
-            .collect(Collectors.<CompletableFuture<SearchResult<T>>>toList());
-
-    // TODO: if all fails return error instead of empty and add test
-
-    // Using the spotify library ( this method is much easier to operate then using
-    // CompletableFuture.allOf and converting the CompletableFuture<Void> to
-    // CompletableFuture<List<SearchResult>>
-    CompletableFuture<List<SearchResult<T>>> searchResultFuture =
-        CompletableFutures.allAsList(queries);
+                    queryExecutorService.submit(
+                        currentTraceContext.wrap(
+                            () -> {
+                              try {
+                                if (Thread.interrupted()) {
+                                  LOG.warn(
+                                      "Chunk query thread timed out without starting work, returning error result.");
+                                  return errorResult;
+                                }
+                                return chunk.query(query);
+                              } catch (Exception err) {
+                                // Only log the exception message as warn, and not the entire trace
+                                // as this can cause performance issues if significant amounts of
+                                // invalid queries are received
+                                LOG.warn("Chunk Query Exception: {}", err.getMessage());
+                                LOG.debug("Chunk Query Exception", err);
+                                // We catch IllegalArgumentException ( and any other exception that
+                                // represents a parse failure ) and instead of returning an empty
+                                // result we throw back an error to the user
+                                if (err instanceof IllegalArgumentException) {
+                                  throw err;
+                                }
+                                return errorResult;
+                              }
+                            })))
+            .peek(
+                (future) ->
```

Comment:
Can we move this cancellation logic into the map? Does this have to be in peek?

Reply:
We need a reference to the returned future in order to schedule a cancellation. You can do this all in a single map, but it ended up being less readable as you need to expand the lambda body and use an intermediate variable. 

You can also do this with multiple map steps (ie, `.map().map().collect()`) but IntelliJ will flag this as unnecessary, as you do not need to mutate the future. You just need the reference in order to schedule the cancellation.